iothub/device/src/ClientPropertyCollection.cs

```diff
@@ -0,0 +1,378 @@
+using Microsoft.Azure.Devices.Shared;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// A collection of properties for the client.
+    /// </summary>
+    public class ClientPropertyCollection : PayloadCollection
+    {
+        private const string VersionName = "$version";
+
+        /// <inheritdoc path="/exception['ArgumentException']" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <summary>
+        /// Adds the value for the collection.
+        /// </summary>
+        /// <remarks>
+        /// If the collection has a key that matches the property name this method will throw an <see cref="ArgumentException"/>.
+        /// <para>
+        /// When using this as part of the writable property flow to respond to a writable property update you should pass in the value
+        /// as an instance of <see cref="PayloadSerializer.CreateWritablePropertyResponse(object, int, long, string)"/>
+        /// to ensure the correct formatting is applied when the object is serialized.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="ArgumentNullException"><paramref name="propertyName"/> is <c>null</c> </exception>
+        /// <param name="propertyName">The name of the property to add.</param>
+        /// <param name="propertyValue">The value of the property to add.</param>
+        public override void Add(string propertyName, object propertyValue)
+            => Add(null, propertyName, propertyValue);
+
+        /// <inheritdoc path="/remarks" cref="Add(string, object)" />
+        /// <inheritdoc path="/seealso" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <inheritdoc path="/exception['ArgumentException']" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <summary>
+        /// Adds the value for the collection.
+        /// </summary>
+        /// <exception cref="ArgumentNullException"><paramref name="propertyName"/> is <c>null</c> </exception>
+        /// <param name="componentName">The component with the property to add.</param>
+        /// <param name="propertyName">The name of the property to add.</param>
+        /// <param name="propertyValue">The value of the property to add.</param>
+        public void Add(string componentName, string propertyName, object propertyValue)
+            => AddInternal(new Dictionary<string, object> { { propertyName, propertyValue } }, componentName, false);
+
+        /// <inheritdoc path="/remarks" cref="Add(string, object)" />
+        /// <inheritdoc path="/seealso" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <inheritdoc path="/exception['ArgumentException']" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <summary>
+        /// Adds the value for the collection.
+        /// </summary>
+        /// <param name="properties">A collection of properties to add or update.</param>
```

Comment:
It seems like some of these methods allow for you to "add" a key that already exists in order to update, but some other methods don't. Is there some reason for that? Is that just how IDictionary works?

Reply:
Looking at ```public void AddOrUpdate(IDictionary<string, object> properties)``` I'm guessing this method is supposed to throw if one or more keys in the added collection already exist? That isn't reflected in this description, if that is the case

Reply:
Correct, we have `Add`, which will use IDictionary.Add => will throw ArgumentException if the key already exists. 
`AddOrUpdate` will force-update the dictionary instead => `Collection[key] = value`.